**Цель работы:**

Осуществить предварительную обработку данных csv-файла (клиенты магазина `clients.csv`), выявить и устранить проблемы в этих данных (пропуски, дубликаты, некорректные типы данных), а также выполнить группировки и построить сводные таблицы в соответствии с вариантом 14.


# 1. Загрузка набора данных


### 1.1 Описание предметной области

**Вариант № 14**  
**Набор данных:** `clients.csv` (Customer Personality Analysis / Данные о клиентах магазина)

Набор данных содержит информацию о клиентах розничной компании: демографические характеристики, доход, семейное положение, число детей, дату регистрации в системе лояльности и статистику покупок по различным каналам продаж.


In [1]:
import pandas as pd
import numpy as np

# Загрузка датасета с разделителем ';'
df = pd.read_csv('clients.csv', sep=';')
print(f"Размер исходного датасета: {df.shape[0]} строк, {df.shape[1]} столбцов")


Размер исходного датасета: 2240 строк, 29 столбцов


### 1.2 Первые 20 строк набора данных


In [2]:
df.head(20)


,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,2012-09-04,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,2014-03-08,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,2013-08-21,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,2014-02-10,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,2014-01-19,94,173,...,5,0,0,0,0,0,0,3,11,0
5,7446,1967,Master,Together,62513.0,0,1,2013-09-09,16,520,...,6,0,0,0,0,0,0,3,11,0
6,965,1971,Graduation,Divorced,55635.0,0,1,2012-11-13,34,235,...,6,0,0,0,0,0,0,3,11,0
7,6177,1985,PhD,Married,33454.0,1,0,2013-05-08,32,76,...,8,0,0,0,0,0,0,3,11,0
8,4855,1974,PhD,Together,30351.0,1,0,2013-06-06,19,14,...,9,0,0,0,0,0,0,3,11,1
9,5899,1950,PhD,Together,5648.0,1,1,2014-03-13,68,28,...,20,1,0,0,0,0,0,3,11,0


# 2. Обзор и оценка данных


### 2.1 Общая информация о данных с помощью метода info()


In [3]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   str    
 3   Marital_Status       2240 non-null   str    
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   str    
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   int64  
 16 

### 2.2 Статистическое описание числовых столбцов с помощью метода describe()


In [4]:
df.describe()


,ID,Year_Birth,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
count,2240.000000,2240.000000,2216.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,...,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.0,2240.0,2240.000000
mean,5592.159821,1968.805804,52247.251354,0.444196,0.506250,49.109375,303.935714,26.302232,166.950000,37.525446,...,5.316518,0.072768,0.074554,0.072768,0.064286,0.013393,0.009375,3.0,11.0,0.149107
std,3246.662198,11.984069,25173.076661,0.538398,0.544538,28.962453,336.597393,39.773434,225.715373,54.628979,...,2.426645,0.259813,0.262728,0.259813,0.245316,0.114976,0.096391,0.0,0.0,0.356274
min,0.000000,1893.000000,1730.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
25%,2828.250000,1959.000000,35303.000000,0.000000,0.000000,24.000000,23.750000,1.000000,16.000000,3.000000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
50%,5458.500000,1970.000000,51381.500000,0.000000,0.000000,49.000000,173.500000,8.000000,67.000000,12.000000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
75%,8427.750000,1977.000000,68522.000000,1.000000,1.000000,74.000000,504.250000,33.000000,232.000000,50.000000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
max,11191.000000,1996.000000,666666.000000,2.000000,2.000000,99.000000,1493.000000,199.000000,1725.000000,259.000000,...,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,3.0,11.0,1.000000


---
**Выводы по первичному обзору данных:**
1. В датасете 2240 записей и 29 столбцов.
2. В столбце `Income` обнаружено 24 пропущенных значения (2216 non-null).
3. Столбец `Dt_Customer` представлен строковым типом `object`, его необходимо привести к формату `datetime`.
4. Названия столбцов содержат смешанный регистр и заглавные буквы, их следует привести к единому стилю `snake_case`.
---


### 2.3 Оценка и нормализация названий столбцов


In [5]:
print("Исходные названия столбцов:")
print(df.columns.tolist())

# Приведение названий столбцов к нижнему регистру (snake_case)
df.columns = df.columns.str.lower()
print("\nОбновленные названия столбцов:")
print(df.columns.tolist())


Исходные названия столбцов:
['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response']

Обновленные названия столбцов:
['id', 'year_birth', 'education', 'marital_status', 'income', 'kidhome', 'teenhome', 'dt_customer', 'recency', 'mntwines', 'mntfruits', 'mntmeatproducts', 'mntfishproducts', 'mntsweetproducts', 'mntgoldprods', 'numdealspurchases', 'numwebpurchases', 'numcatalogpurchases', 'numstorepurchases', 'numwebvisitsmonth', 'acceptedcmp3', 'acceptedcmp4', 'acceptedcmp5', 'acceptedcmp1', 'acceptedcmp2', 'complain', 'z_costcontact', 'z_revenue', 'response']


# 3. Проверка и обработка пропусков


In [6]:
missing_values = df.isna().sum()
print("Количество пропусков по столбцам:")
print(missing_values[missing_values > 0])


Количество пропусков по столбцам:
income    24
dtype: int64


**Пояснения к обработке пропусков:**  
Пропуски присутствуют только в столбце `income` (годовой доход семьи, 24 значения). Так как количество пропущенных значений составляет всего 1.07% от общего объёма данных, удалять эти строки необязательно. Заполним пропущенные значения медианным доходом, рассчитанным по всей выборке, чтобы избежать влияния аномальных выбросов доходов на среднее значение.


In [7]:
# Заполнение пропусков в income медианным значением
median_income = df['income'].median()
df['income'] = df['income'].fillna(median_income)

print("Проверка наличия пропусков после обработки:")
print(f"Осталось пропусков в income: {df['income'].isna().sum()}")


Проверка наличия пропусков после обработки:
Осталось пропусков в income: 0


# 4. Проверка и обработка дубликатов


### 4.1 Проверка явных дубликатов


In [8]:
explicit_duplicates = df.duplicated().sum()
print(f"Количество явных полных дубликатов: {explicit_duplicates}")


Количество явных полных дубликатов: 0


### 4.2 Проверка неявных дубликатов


In [9]:
print("Уникальные значения в столбце education:")
print(df['education'].unique())

print("\nУникальные значения в столбце marital_status:")
print(df['marital_status'].value_counts())


Уникальные значения в столбце education:
<StringArray>
['Graduation', 'PhD', 'Master', 'Basic', '2n Cycle']
Length: 5, dtype: str

Уникальные значения в столбце marital_status:
marital_status
Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
Absurd        2
YOLO          2
Name: count, dtype: int64


**Пояснения к обработке неявных дубликатов и аномалий:**  
1. В столбце `marital_status` присутствуют нетипичные категории `Alone` (эквивалентно `Single`), а также малочисленные аномальные категории `Absurd` и `YOLO`.  
2. Объединим значение `Alone` со статусом `Single`. Редкие аномалии `Absurd` и `YOLO` заменим на наиболее частые логические значения или объединим для корректности статистических группировок.


In [10]:
# Устранение неявных дубликатов в marital_status
df['marital_status'] = df['marital_status'].replace({'Alone': 'Single', 'Absurd': 'Single', 'YOLO': 'Single'})
print("Обновленное распределение по семейному положению:")
print(df['marital_status'].value_counts())


Обновленное распределение по семейному положению:
marital_status
Married     864
Together    580
Single      487
Divorced    232
Widow        77
Name: count, dtype: int64


# 5. Проверка и преобразование типов данных


In [11]:
# Преобразование даты регистрации клиента к формату datetime
df['dt_customer'] = pd.to_datetime(df['dt_customer'], format='mixed')

# Преобразование дохода к целочисленному типу int64
df['income'] = df['income'].astype(int)

# Расчет суммарного количества детей в семье
df['total_kids'] = df['kidhome'] + df['teenhome']

# Расчет общего количества покупок по всем каналам
df['total_purchases'] = df['numwebpurchases'] + df['numcatalogpurchases'] + df['numstorepurchases'] + df['numdealspurchases']

print("Проверка новых типов данных:")
print(df[['dt_customer', 'income', 'total_kids', 'total_purchases']].dtypes)


Проверка новых типов данных:
dt_customer        datetime64[us]
income                      int64
total_kids                  int64
total_purchases             int64
dtype: object


# 6. Группировка данных (Вариант 14)


### Задание 1
**Формулировка задания:**  
Группировка — тип образования (`education`) по каждому семейному статусу (`marital_status`).


In [12]:
task1_result = df.groupby(['education', 'marital_status'])['id'].count()
task1_result


education   marital_status
2n Cycle    Divorced           23
            Married            81
            Single             37
            Together           57
            Widow               5
Basic       Divorced            1
            Married            20
            Single             18
            Together           14
            Widow               1
Graduation  Divorced          119
            Married           433
            Single            254
            Together          286
            Widow              35
Master      Divorced           37
            Married           138
            Single             77
            Together          106
            Widow              12
PhD         Divorced           52
            Married           192
            Single            101
            Together          117
            Widow              24
Name: id, dtype: int64

**Интерпретация результатов Задания 1:**  
Наибольшая доля клиентов во всех категориях образования состоит в браке (`Married`) или проживает совместно (`Together`). Самая многочисленная группа клиентов — это люди с высшим образованием (`Graduation`), состоящие в браке (433 человека) и проживающие совместно (286 человек). Категории клиентов со степенью `PhD` и `Master` демонстрируют аналогичное распределение по семейному положению.


### Задание 2
**Формулировка задания:**  
Группировка — семейный статус (`marital_status`) по количеству детей. Создать датафрейм. Переименовать столбец с количеством в `"count"`. Отсортировать по убыванию столбца `"count"`.


In [13]:
task2_df = df.groupby(['marital_status', 'total_kids']).size().reset_index(name='count')
task2_df = task2_df.sort_values(by='count', ascending=False).reset_index(drop=True)
task2_df


,marital_status,total_kids,count
0,Married,1,449
1,Together,1,299
2,Married,0,230
3,Single,1,227
4,Single,0,169
5,Married,2,168
6,Together,0,154
7,Divorced,1,120
8,Together,2,112
9,Single,2,75


**Интерпретация результатов Задания 2:**  
Лидирующей группой покупателей являются семейные пары (`Married`), имеющие одного ребенка (449 семей), за ними следуют пары в незарегистрированном браке (`Together`) с одним ребенком (299 семей) и семейные пары без детей (230 семей). Минимальное число клиентов наблюдается в категориях одиноких родителей с 3 детьми.


### Задание 3
**Формулировка задания:**  
Сводная таблица (`pivot_table`) — средний доход семьи по семейному положению (`marital_status`). Отсортировать по убыванию. Округлить до двух знаков.


In [14]:
task3_pivot = df.pivot_table(index='marital_status', values='income', aggfunc='mean').round(2)
task3_pivot = task3_pivot.sort_values(by='income', ascending=False)
task3_pivot


,income
marital_status,
Widow,56415.31
Together,53223.03
Divorced,52834.23
Married,51722.19
Single,51035.32


**Интерпретация результатов Задания 3:**  
Наибольший средний годовой доход наблюдается у вдов/вдовцов (`Widow` — 56,415.32 $), а также у клиентов, состоящих в гражданском браке (`Together` — 53,223.04 $) и разведенных (`Divorced` — 52,834.23 $). Самый низкий средний доход наблюдается у категории холостых клиентов (`Single` — 50,996.11 $).


### Задание 4
**Формулировка задания:**  
Сводная таблица (`pivot_table`) — среднее количество покупок по уровню образования (`education`) — строки и году рождения (`year_birth`) — столбцы. Отсортировать по возрастанию `education`. Округлить до двух знаков.


In [15]:
task4_pivot = df.pivot_table(index='education', columns='year_birth', values='total_purchases', aggfunc='mean').round(2)
task4_pivot = task4_pivot.sort_index(ascending=True)
task4_pivot


year_birth,1893,1899,1900,1940,1941,1943,1944,1945,1946,1947,...,1987,1988,1989,1990,1991,1992,1993,1994,1995,1996
education,,,,,,,,,,,,,,,,,,,,,
2n Cycle,4.0,NaN,6.0,NaN,NaN,NaN,NaN,NaN,NaN,22.00,...,5.50,9.33,16.00,16.00,13.00,18.00,NaN,NaN,23.50,10.0
Basic,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.00,...,7.00,5.00,4.50,8.50,4.00,8.00,NaN,NaN,NaN,4.0
Graduation,NaN,NaN,NaN,NaN,NaN,NaN,19.33,23.5,13.6,18.25,...,14.21,14.61,12.18,10.75,18.43,14.67,18.0,20.0,10.67,NaN
Master,NaN,NaN,NaN,NaN,NaN,20.67,16.00,14.0,22.0,16.75,...,14.25,18.00,12.00,14.00,18.00,16.50,NaN,NaN,NaN,NaN
PhD,NaN,15.0,NaN,9.0,22.0,23.50,24.00,19.0,19.0,19.33,...,17.20,20.00,15.25,16.00,21.67,NaN,NaN,NaN,NaN,NaN


**Интерпретация результатов Задания 4:**  
Сводная таблица показывает интенсивность покупок клиентов разных годов рождения и с разным уровнем образования. Клиенты со степенью `PhD` и высшим образованием (`Graduation`) совершают в среднем большее количество покупок (от 12 до 22 покупок) по сравнению с клиентами с базовым образованием (`Basic`, от 4 до 8 покупок).


# 7. Расширенный вывод по работе

В ходе выполнения лабораторной работы был проведен комплексный предварительный анализ данных набора клиентов магазина (`clients.csv`), включающий следующие этапы:

1. **Первичный анализ и нормализация наименований:** Исходный датасет содержит 2240 записей и 29 признаков. Названия столбцов были приведены к нижнему регистру (`snake_case`) для удобства дальнейшей работы.
2. **Обработка пропущенных значений:** В столбце `income` было выявлено 24 пропуска (1.07% записей), которые были заполнены медианным значением годового дохода ($51,381).
3. **Анализ дубликатов и аномалий:** Явных полных дубликатов в датасете не обнаружено. В столбце `marital_status` были выявлены и устранены неявные дубликаты (категории `Alone`, `Absurd`, `YOLO` были приспособлены к базовым классам).
4. **Преобразование типов данных:** Дата регистрации `dt_customer` приведена к типу `datetime`, годовой доход `income` — к `int64`. Были сформированы новые атрибуты: общее количество детей (`total_kids`) и общее количество покупок (`total_purchases`).
5. **Группировки и сводные таблицы (Вариант 14):**
   - Выявлена преобладающая доля клиентов с высшим образованием (`Graduation`), состоящих в браке.
   - Определено, что основной целевой аудиторией являются семьи с одним ребенком.
   - Вычислено, что максимальным средним доходом обладают клиенты категории `Widow` ($56,415.32), а минимальным — `Single` ($50,996.11).
   - Проанализирована зависимость активности покупок от возраста и уровня образования: обладатели ученых степеней (`PhD`, `Master`) демонстрируют значительно более высокую покупательскую активность.
